### DISCUSSION TITLE
How should 3rd party GraphQL client work with Quarkus Smallrye GraphQL subscription?

### DISCUSSION THREAD

**Discussion by tigerinus** [quarkusio/quarkus#25967] [Q&A] [answered]
I am very new to Quarkus + GraphQL.

I built a GraphQL service just for learning purpose at https://github.com/tigerinus/quakus-graphql-demo

```java
import java.util.Collection;

import javax.inject.Inject;

import org.eclipse.microprofile.graphql.GraphQLApi;
import org.eclipse.microprofile.graphql.Mutation;
import org.eclipse.microprofile.graphql.Query;

import com.wangxiaohu.quarkus.graphql.demo.model.Person;
import com.wangxiaohu.quarkus.graphql.demo.service.PersonService;

import io.quarkus.logging.Log;
import io.smallrye.graphql.api.Subscription;
import io.smallrye.mutiny.Multi;
import io.smallrye.mutiny.operators.multi.processors.BroadcastProcessor;

@GraphQLApi
public class PersonResource {

    private final BroadcastProcessor<Person> _personBroadcastProcessor;

    public PersonResource() {
        _personBroadcastProcessor = BroadcastProcessor.create();
    }

    @Inject
    PersonService _personService;

    @Query("getAllPeople")
    public Collection<Person> getAllPeople() {
        return _personService.getAllPeople();
    }

    @Query("getPersonById")
    public Person getPerson(int id) {
        return _personService.getPerson(id);
    }

    @Mutation("createPerson")
    public Person createPerson(String firstName, String lastName) {
        Person person = _personService.createPerson(firstName, lastName);

        Log.info("signaling the person created...");
        _personBroadcastProcessor.onNext(person);
        Log.info("signaled the person created.");

        return person;
    }

    @Subscription("personCreated")
    public Multi<Person> subscribeToPersonCreation() {
        Log.info("subscribeToPersonCreation");
        return _personBroadcastProcessor;
    }
}
```

The service allows adding person, getting all people, and subscription to person creation.

I also built a small test code in Python to subscribe to the person creation at https://github.com/tigerinus/quakus-graphql-demo/tree/master/test/python

```python
from gql import gql, Client
from gql.transport.websockets import WebsocketsTransport

if __name__ == '__main__':
    transport = WebsocketsTransport(
        url="ws://localhost:8080/graphql",
        subprotocols=[WebsocketsTransport.GRAPHQLWS_SUBPROTOCOL]
    )

    client = Client(transport=transport, fetch_schema_from_transport=True)

    query = gql(
        '''
        subscription subscribeToPersonCreation {
            personCreated{
                id
                firstName
                lastName
            }
        }
        '''
    )

    for result in client.subscribe(query):
        print(result)
```

However, adding a new person does not trigger the subscription for some reason.

I've set a breakpoint at `_personBroadcastProcessor.onNext(person)` method, then I see `subscribers` is empty

![image](https://user-images.githubusercontent.com/7172560/172075549-3fb1f083-27d8-4efd-9d18-1d1c93a8bef6.png)

Can someone let me know what am I missing here?

Thanks!

[a comment by an automated account is not shown]

**Comment by tigerinus** [comment]
I also tried writting a GraphQL client in nodejs to subscribe, but not receiving anything either upon creating a new `person` record:

```js
const ws = require('ws');
const Crypto = require('crypto');
const { createClient } = require('graphql-ws');

const client = createClient({
  url: "ws://localhost:8080/graphql",
  webSocketImpl: ws,
  generateID: () =>
    ([1e7] + -1e3 + -4e3 + -8e3 + -1e11).replace(/[018]/g, (c) =>
      (c ^ (Crypto.randomBytes(1)[0] & (15 >> (c / 4)))).toString(16),
    ),
});

(async () => {
  const onNext = (value) => {
    console.log(value);
  }

  await new Promise((resolve, reject) => {
    unsubscribe = client.subscribe(
      {
        query: `subscription subscribeToPersonCreation {
          personCreated {
            id
            firstName
            lastName
          }
        }`
      },
      {
        next: onNext,
        error: reject,
        complete: resolve,
      }
    );
  });
})();
```

See actual code at https://github.com/tigerinus/quakus-graphql-demo/tree/master/test/nodejs

**Reply by phillip-kruger** [reply]
Does the subscription work when you use the built in UI (graphiql)

**Reply by tigerinus** [reply]
> Does the subscription work when you use the built in UI (graphiql)

I am very new to this GraphQL UI. How exactly do I check the subscription?

If I run 

```graphql
subscription personCreated {
  personCreated {
    id
    firstName
    lastName
  }
}
```

I get

```graphql
{
  "data": {
    "upstreamPublisher": {}
  }
}
```

**Reply by phillip-kruger** [reply]
That seems ok, it's waiting for data to stream. Do you have a reproducer I can look at ?

**Reply by phillip-kruger** [reply]
Ok I see you have a repo linked, @jmartisk will have a look

**Reply by jmartisk** [reply]
Yeah but sorry I won't get round to it this week, I'm abroad on a conference

**Reply by tigerinus** [reply]
While waiting for @jmartisk to take a look, I took another look at the runtime callstack and here are findings:

1. Breakpoints

- set a breakpoint after `addSubscription(subscription)` in `BroadcastProcessor::subscribe` method 
- set a breakpoint at `_personBroadcastProcessor.onNext(person);` in my own code

2. Run the service

3. Run `subscription` in one GraphQL UI

- breakpoint at `BroadcastProcessor::subscribe`  is hit
- `subscribers` of `BroadcastProcessor` object  has the new subscriber - great!
- The `BroadcastProcessor` object address is `@11034`
- continue

4. Run `mutation` in another GraphQL UI

- breakpoint at `_personBroadcastProcessor.onNext(person);` is hit
-  `subscribers` of `BroadcastProcessor` object is still empty 
   - *this is the interesting part.*
- The `BroadcastProcessor` object address is `@11044`
  - *different from the object in step 3 above!*

5. Look at the log

```
2022-06-11 22:30:00,333 INFO  [com.wan.qua.gra.dem.PersonResource] (vert.x-eventloop-thread-0) subscribeToPersonCreation
2022-06-11 22:37:15,222 INFO  [com.wan.qua.gra.dem.PersonResource] (vert.x-worker-thread-1) signaling the person created...
```

Looks like subscription and `onNext(...)` happened in different threads...

Therefore, I had to use following technique to ensure the same BroadcastProcessor is used:

```java
@ApplicationScoped
public class PersonBroadcastProcessor {

    private static final BroadcastProcessor<Person> PERSON_BROADCAST_PROCESSOR = BroadcastProcessor.create();

    @Produces
    BroadcastProcessor<Person> personBroadcastProcessor() {
        return PERSON_BROADCAST_PROCESSOR;
    }
}
```

This solves the problem. 

Is this the right way to do it, or should [the guide](https://quarkus.io/guides/smallrye-graphql) be updated?

Thanks.

**Reply by jmartisk** [reply]
Sorry for the delay, I was finally able to  have a look.
The problem is that a `GraphQLApi` bean, unless you define its CDI scope, is request-scoped. So if you have a non-static `BroadcastProcessor` field  in it, then the `BroadcastProcessor` returned from the `@Subscription` method will be a different instance than the `BroadcastProcessor` to which you are submitting events while handling subsequent requests later, that's why the events are not processed. 
To fix this, you can either create the `@ApplicationScoped` processor producer as you did, or, the easier solution is to make the `GraphQLApi` class application-scoped.

We have the `@ApplicationScoped` annotation in the quickstart https://github.com/quarkusio/quarkus-quickstarts/blob/2.9.2.Final/microprofile-graphql-quickstart/src/main/java/org/acme/microprofile/graphql/FilmResource.java#L19 but it seems that we might want to add that into the guide too... It indeed isn't clear from the guide that this is necessary. I'll send a PR to update it

**Reply by tigerinus** [reply]
I would say making broadcast processor `@ApplicationScoped` gives better scalability than making `GraphQLApi` `@ApplicationScoped`, would you agree?

You don't want to have thousands requests handled by one thread - the `FilmResource` in your example. It'd be nice for the guide to point this out.

**Reply by tigerinus** [reply]
The guide should also point out even with broadcast processor `@ApplicationScoped`, it still wouldn't work if there are multiple pod instances of the service across K8s nodes. In that case people would have to rely on some distributed storage with pub/sub support instead of broadcast processor, e.g. Redis, Kafka, etc.

**Comment by tigerinus** [comment, marked as answer]
This is answered in the reply thread.

**Reply by phillip-kruger** [reply]
@jmartisk ,the default is application scope a.f.a.i.k

**Reply by jmartisk** [reply]
I looked in the docs and default is actually `@Dependent` (note at the end of https://quarkus.io/guides/cdi-integration#additional_bean_build_item), which for a HTTP-based endpoint is basically the same as request-scoped.

**Reply by phillip-kruger** [reply]
O ok, I wonder if we should not change that to application scope

**Reply by jmartisk** [reply]
Hmm, we might actually do it, except rather go with `Singleton`. JAX-RS endpoints are Singleton by default.
I reported https://github.com/quarkusio/quarkus/issues/26103

**Reply by phillip-kruger** [reply]
Great. will you look into a PR for this ?

**Reply by jmartisk** [reply]
Of course

**Reply by tigerinus** [reply]
> Hmm, we might actually do it, except rather go with `Singleton`. JAX-RS endpoints are Singleton by default. I reported #26103

My concern is this does not scale in real world scenario but only helps demonstration. It should be how it was, then the guide should educate developers how to get multiple GraphQL endpoints to share a broadcast object, whether by making it singleton or using a distributed storage that supports pub/sub.

**Reply by phillip-kruger** [reply]
@jmartisk - I did not check, but if a user adds `@RequestScope`will that override the now default `@Singleton` ? If not, it should

**Reply by jmartisk** [reply]
> @jmartisk - I did not check, but if a user adds `@RequestScope`will that override the now default `@Singleton` ? If not, it should

Of course, we only changed the default (what happens if you don't add a scope annotation).

> My concern is this does not scale in real world scenario but only helps demonstration. It should be how it was, then the guide should educate developers how to get multiple GraphQL endpoints to share a broadcast object, whether by making it singleton or using a distributed storage that supports pub/sub.

True, but I think that building a distributed architecture where events are shared between service instances (through Kafka or similar) is beyond the scope of the Getting Started with GraphQL guide. We might turn that into a blog post or something. 

For now, we've aligned the default scope to be the same as with JAX-RS endpoints, because having JAX-RS and GraphQL endpoints behave differently leads to confusion. One would quite reasonably expect that a `BroadcastProcessor` instance inside the API and returned by a `@Subscription` method will be able to broadcast events to all subscriptions, just like the exact same scenario would work with a JAX-RS endpoint that produces server-sent events.

**Reply by jmartisk** [reply]
@tigerinus I need to add; even though I think describing a lot of details about building distributed architectures doesn't belong in the guide, if you have an idea how to gently improve it, PRs are welcome, of course!

**Reply by phillip-kruger** [reply]
I also think `@ApplicationScope` or `@Singleton` is fine for a default. If users want they can scope the Bean different if needed.

**Reply by tigerinus** [reply]
> @tigerinus I need to add; even though I think describing a lot of details about building distributed architectures doesn't belong in the guide, if you have an idea how to gently improve it, PRs are welcome, of course!

Agreed the guide should not include a lot of details.

If the guide could just point out that BroadcastProcessor should be shared across GraphQL endpoint threads, I would immediately understand and wouldn't run into this discussion.

Along with this, the guide could just add one extra message like "for scalability in production environment, it's recommended to replace it with a distributed storage that supports pub/sub model such as Redis, Kafka. However this is beyond the scope of this guide".

I will also update my sample project backed with Redis or Kafka and leave people to get there via search engine to learn how to do it.
